src/software/Localization/main_SfM_Localization_MultipleImages.cpp

```diff
@@ -0,0 +1,341 @@
+      {
+        std::cerr << "Cannot open the input provided image : " << *iter_image << std::endl;
+        return EXIT_FAILURE;
+      }
+      const std::string
+        sFeat = stlplus::create_filespec(sMatchesDir, stlplus::basename_part(sView_filename.c_str()), "feat"),
+        sDesc = stlplus::create_filespec(sMatchesDir, stlplus::basename_part(sView_filename.c_str()), "desc");
+
+      // Compute features and descriptors and save them if they don't exist yet
+      if (!stlplus::file_exists(sFeat) || !stlplus::file_exists(sDesc))
+      {
+        image_describer->Describe(imageGray, query_regions);
+        image_describer->Save(query_regions.get(), sFeat, sDesc);
+        std::cout << "#regions detected in query image: " << query_regions->RegionCount() << std::endl;
+      }
+      else // load already existing regions
+      {
+        query_regions->Load(sFeat,sDesc);
+      }
+    }
+
+    std::shared_ptr<cameras::IntrinsicBase> optional_intrinsic(nullptr);
+    if (cmd.used('s') && num_initial_intrinsics == 1)
+    {
+      optional_intrinsic = std::make_shared<cameras::Pinhole_Intrinsic_Radial_K3>(
+        imageGray.Width(), imageGray.Height(),
+        ptrPinhole->focal(), ptrPinhole->principal_point()[0], ptrPinhole->principal_point()[1],
+        ptrPinhole->getParams()[3], ptrPinhole->getParams()[4], ptrPinhole->getParams()[5]);
+    }
+
+    geometry::Pose3 pose;
+    sfm::Image_Localizer_Match_Data matching_data;
+    matching_data.error_max = dMaxResidualError;
+
+    // Build the view corresponding to the image
+    View v(*iter_image, views.size(), views.size(), views.size(), imageGray.Width(), imageGray.Height());
+
+    // Try to localize the image in the database thanks to its regions
+    if (!localizer.Localize(
+      Pair(imageGray.Width(), imageGray.Height()),
+      optional_intrinsic.get(),
+      *(query_regions.get()),
+      pose,
+      &matching_data))
+    {
+      std::cerr << "Cannot locate the image " << *iter_image << std::endl;
+      v.id_intrinsic = UndefinedIndexT;
+      v.id_pose = UndefinedIndexT;
+    }
+    else
+    {
+      const bool b_new_intrinsic = (optional_intrinsic == nullptr);
+      // A valid pose has been found (try to refine it):
+      // If not intrinsic as input:
+      // init a new one from the projection matrix decomposition
+      // Else use the existing one and consider as static.
+      if (b_new_intrinsic)
+      {
+        // setup a default camera model from the found projection matrix
+        Mat3 K, R;
+        Vec3 t;
+        KRt_From_P(matching_data.projection_matrix, &K, &R, &t);
+
+        const double focal = (K(0,0) + K(1,1))/2.0;
+        const Vec2 principal_point(K(0,2), K(1,2));
+        optional_intrinsic = std::make_shared<cameras::Pinhole_Intrinsic_Radial_K3>(
+          imageGray.Width(), imageGray.Height(),
+          focal, principal_point(0), principal_point(1));
+
+      }
+      sfm::SfM_Localizer::RefinePose
```

Comment:
we must check the return boolean value, BA can failed to optimize the poses for an unknown reason...